sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,326 @@
+import numpy as np
+
+from ..base import MetaEstimatorMixin, clone
+from ..utils.validation import check_X_y, check_array, check_is_fitted
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils import safe_mask
+from ..exceptions import ConvergenceWarning
+from ..utils.metaestimators import _BaseComposition
+import warnings
+
+__all__ = ["SelfTrainingClassifier"]
+
+# Authors: Oliver Rausch   <rauscho@ethz.ch>
+#          Patrice Becker  <beckerp@ethz.ch>
+# License: BSD 3 clause
+
+
+def _validate_estimator(estimator):
+    """Make sure that an estimator implements the necessary methods."""
+    if not hasattr(estimator, "predict_proba"):
+        msg = "base_classifier ({}) should implement predict_proba!"
+        raise ValueError(msg.format(type(estimator).__name__))
+
+
+class SelfTrainingClassifier(MetaEstimatorMixin, _BaseComposition):
+    """Self-training classifier
+
+    This class allows a given supervised classifier to function as a
+    semi-supervised classifier, allowing it to learn from unlabeled data. It
+    does this by iteratively predicting labels for the unlabeled data.
+
+    Read more in the :ref:`User Guide <self_training>`.
+
+    Parameters
+    ----------
+    base_classifier : estimator object
+        An estimator object implementing ``fit`` and ``predict_proba``.
+        Invoking the ``fit`` method will fit a clone of the passed estimator,
+        which will be stored in the ``base_classifier_`` attribute.
+
+    threshold : float, optional (default=0.75)
+        The decision threshold. If the ``base_classifier`` makes a prediction
+        with a ``predict_proba`` above this threshold, it will be added to the
+        labeled dataset.
+        Should be in [0, 1).
```

Comment:
I don't know if it's worth noting that the usefulness of the threshold will rely on the estimator being well calibrated. Surely there are versions of self training that just take a top %ile of the data at each iteration? I've not looked into the papers...

Reply:
>I don't know if it's worth noting that the usefulness of the threshold will rely on the estimator being well calibrated.

Good point. [These](https://link.springer.com/article/10.1007/s13042-015-0328-7) guys study that at least for trees.

Reply:
I've added the `n_best` `selection_criterion` that enables this.